crypto/hpke/hpke.c

```diff
@@ -0,0 +1,2178 @@
+ * @param pub_len is the length of pub (buffer size on input)
+ * @param cipher is a random value of the appropriate length for a ciphertext
+ * @param cipher_len is the length of cipher
+ * @return 1 for success, otherwise failure
+ */
+int OSSL_HPKE_get_grease_value(OSSL_LIB_CTX *libctx, const char *propq,
+                               OSSL_HPKE_SUITE *suite_in,
+                               OSSL_HPKE_SUITE *suite,
+                               unsigned char *pub,
+                               size_t *pub_len,
+                               unsigned char *cipher,
+                               size_t cipher_len)
+{
+    return hpke_get_grease_value(libctx, propq, suite_in, suite,
+                                 pub, pub_len, cipher, cipher_len);
+}
+
+/*
+ * @brief map a string to a HPKE suite
+ * @param str is the string value
+ * @param suite is the resulting suite
+ * @return 1 for success, otherwise failure
+ */
+int OSSL_HPKE_str2suite(const char *str, OSSL_HPKE_SUITE *suite)
+{
+    return hpke_str2suite(str, suite);
+}
+
+/*
+ * @brief tell the caller how big the cipertext will be
+ * @param suite is the suite to be used
+ * @param clearlen is the length of plaintext
+ * @return the length of the related ciphertext or zero on error
+ *
+ * AEAD algorithms add a tag for data authentication.
+ * Those are almost always, but not always, 16 octets
+ * long, and who know what'll be true in the future.
+ * So this function allows a caller to find out how
+ * much data expansion they'll see with a given
+ * suite.
+ */
+size_t OSSL_HPKE_get_ciphertext_size(OSSL_HPKE_SUITE suite, size_t clearlen)
+{
+    size_t enclen = 0;
+    size_t cipherlen = 0;
+
+    if (hpke_expansion(suite, &enclen, clearlen, &cipherlen) != 1)
+        return 0;
+    return cipherlen;
+}
+
+/*
+ * @brief tell the caller how big the public value ``enc`` will be
+ * @param suite is the suite to be used
+ * @return size of public encap or zero on error
+ *
+ * AEAD algorithms add a tag for data authentication.
+ * Those are almost always, but not always, 16 octets
+ * long, and who know what'll be true in the future.
+ * So this function allows a caller to find out how
+ * much data expansion they'll see with a given
+ * suite.
+ */
+size_t OSSL_HPKE_get_public_encap_size(OSSL_HPKE_SUITE suite)
+{
+    size_t enclen = 0;
+    size_t cipherlen = 0;
+    size_t clearlen = 16;
+
+    if (hpke_expansion(suite, &enclen, clearlen, &cipherlen) != 1)
+        return 0;
+    return enclen;
+}
+
+/**
+ * @brief recommend an IKM size in octets for a given suite
+ * @param suite is the suite to be used
+ * @return the recommended size or zero on error
+ *
+ * Today, this really only uses the KEM to recommend
+ * the number of random octets to use based on the
+ * size of a private value. In future, it could also
+ * factor in e.g. the AEAD.
+ */
+size_t OSSL_HPKE_recommend_ikmelen(OSSL_HPKE_SUITE suite)
```

Comment:
I thought the ikm was mainly for testing.. This public API  tends to suggest it is not.

Reply:
wrt testing - the deterministic key gen stuff is part of RFC9180 so I guess should be supported even if only currently used for test vectors.
I'm fine to delete the last comment but not sure I grok what makes it unhelpful. (It's true it's fairly speculative - there'd probably have to be some KEM that had length options that were tied to the kdf_id or aead_id before anything would use more than the kem_id)
I also don't get the "another API" thing.
If it helped for some reason, we could change the input from a suite to just a kem_id if that's felt better.

Reply:
ongoing discussion below so marking this "resolved" to not have two places to type:-)